pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+	// back to the remote node.
+	ctx := logtags.AddTag(r.ctx, "id", tracingID(e.TraceID, e.SpanID))
+	t := traceValue{traceID: e.TraceID, spanID: e.SpanID, ctx: ctx, baseCtx: nil}
+	t.logf(1, "start trace for index %d from remote", e.Index)
+	r.m.Store(e.Index, &t)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	if !r.maybeRegister() {
+		return
+	}
+	span := tracing.SpanFromContext(baseCtx.Context())
+	ctx := logtags.AddTag(r.ctx, "id", tracingID(span.TraceID(), span.SpanID()))
+	// NB: We grab the trace and span id now because the baseCtx.Context may be
+	// nil'ed later.
+	t := traceValue{traceID: span.TraceID(), spanID: span.SpanID(), ctx: ctx, baseCtx: baseCtx}
+
+	// Log the identical entry to both to make it easier to match up later.
+	t.logf(1, "registering entry index %d with tracing id (%s)", ent.Index, tracingID(t.traceID, t.spanID))
+	r.m.Store(kvpb.RaftIndex(ent.Index), &t)
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces on the store.
+	if r.numRegistered.Load() == 0 {
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// FlushAll will unregister all the currently active traces. It is safe to call
+// multiple times, but should always be called when the replica is destroyed.
+func (r *RaftTracer) FlushAll() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.m.Delete(index)
+		r.numRegistered.Add(-1)
+		t.logf(2, "cleanup log index (i%d) during close", index)
+		return true
+	})
+}
+
+// traceIfCovered will log the message if it touches any of the registered trace
+// points. Additionally it returns any saved contexts by index for sending to
+// remote nodes. This typically applies to messages that the leader sends to the
+// followers.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []kvserverpb.TracedEntry {
+	var tracedEntries []kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		if !inRange(index, m.Entries) {
+			return true
+		}
+		tracedEntries = append(tracedEntries,
+			kvserverpb.TracedEntry{
+				Index:   index,
+				TraceID: t.traceID,
+				SpanID:  t.spanID,
+			},
+		)
+		// TODO(baptist): Not all the fields are relevant to log for all
+		// message types. Consider cleaning up what is logged.
+		t.logf(4,
```

Comment:
Consider also printing the range of `m.Entries` indices.

Reply:
Added